bmSettings.lua

```diff
@@ -34,6 +38,369 @@ BMSettings.Constants.UpdateRates = {
+        button_cmd, button_evaluate_label, button_show_label, button_icon,
+        button_icon_type, button_icon_lua, button_timer_type, button_cooldown
+    ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)]],
+        id, buttonData.Label or "", buttonData.highestRenderTime or 0, buttonData.TextColorRGB or "", buttonData.ButtonColorRGB or "",
+        buttonData.CachedCountDown or 0, buttonData.CachedCoolDownTimer or 0, buttonData.CachedToggleLocked or 0,
+        buttonData.CachedLastRan or 0, buttonData.labelMidX or 0, buttonData.labelMidY or 0, buttonData.CachedLabel or "",
+        buttonData.Cmd or "", buttonData.EvaluateLabel and 1 or 0, showLabel and 1 or 0, icon,
+        buttonData.IconType or "Item", buttonData.IconLua or "", buttonData.TimerType or "", buttonData.Cooldown or ""
+    )
+    db:close()
+end
+
+function BMSettings:updateSetDB(setName, buttonNumber, buttonID)
+    local db = self:InitializeDB()
+
+    self:saveToDB(db, "INSERT OR REPLACE INTO sets (set_name, button_number, button_id) VALUES (?, ?, ?)",
+        setName, buttonNumber, buttonID)
+    db:close()
+end
+
+function BMSettings:updateCharacterDB(charName, charData)
+    local db = self:InitializeDB()
+
+    self:saveToDB(db, "INSERT INTO characters (character, character_locked, character_hide_title) VALUES (?, ?, ?)",
+        charName, charData.Locked and 1 or 0, charData.HideTitleBar and 1 or 0)
+
+    if charData.Windows then
+        for windowID, windowData in ipairs(charData.Windows or {}) do
+            windowData.Pos = windowData.Pos or { x = 0, y = 0, } -- Default position
+            for setIndex, setName in ipairs(windowData.Sets or {}) do
+                self:saveToDB(db, [[
+                    INSERT OR REPLACE INTO windows (
+                        server, character, window_id, window_fps, window_button_size, window_advtooltip,
+                        window_compact, window_hide_title, window_width, window_height, window_x, window_y,
+                        window_visible, window_font_size, window_locked, window_theme, window_set_id, window_set_name
+                    ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)]],
+                    mq.TLO.EverQuest.Server(), charName, windowID, windowData.FPS or 0, windowData.ButtonSize or 0, windowData.AdvTooltips and 1 or 0,
+                    windowData.CompactMode and 1 or 0, windowData.HideTitleBar and 1 or 0, windowData.Width or 0, windowData.Height or 0,
+                    windowData.Pos.x or 0, windowData.Pos.y or 0, windowData.Visible and 1 or 0, windowData.Font or 0,
+                    windowData.Locked and 1 or 0, windowData.Theme or "", setIndex, setName
+                )
+            end
+        end
+    end
+    db:close()
+end
+
+function BMSettings:deleteButtonFromDB(id)
+    local db = self:InitializeDB()
+    self:saveToDB(db, "DELETE FROM buttons WHERE button_number = ?", id)

```

Comment:
you are doing operations with WHERE clauses that are not keys still will result in a very slow full table scan.